examples/voice-assistant/assistant_calls.py

```diff
@@ -0,0 +1,117 @@
+import asyncio
+import logging
+from enum import Enum
+from typing import Annotated
+
+from livekit import rtc
+from livekit.agents import (
+    JobContext,
+    JobRequest,
+    WorkerOptions,
+    cli,
+    llm,
+)
+from livekit.agents.voice_assistant import AssistantContext, VoiceAssistant
+from livekit.plugins import deepgram, elevenlabs, openai, silero
+
+
+class Room(Enum):
+    BEDROOM = "bedroom"
+    LIVING_ROOM = "living room"
+    KITCHEN = "kitchen"
+    BATHROOM = "bathroom"
+    OFFICE = "office"
+
+
+class AssistantFnc(llm.FunctionContext):
+    @llm.ai_callable(desc="Turn on/off the lights in a room")
+    async def toggle_light(
+        self,
+        room: Annotated[Room, llm.TypeInfo(desc="The specific room")],
+        status: bool,
+    ):
+        logging.info("toggle_light %s %s", room, status)
+        ctx = AssistantContext.get_current()
+        key = "enabled_rooms" if status else "disabled_rooms"
+        li = ctx.get_metadata(key, [])
+        li.append(room)
+        ctx.store_metadata(key, li)
+
+    @llm.ai_callable(desc="User want the assistant to stop/pause speaking")
+    def stop_speaking(self):
+        pass  # do nothing
+
+
+async def entrypoint(ctx: JobContext):
+    gpt = openai.LLM(model="gpt-4-turbo")
+
+    assistant = VoiceAssistant(
+        vad=silero.VAD(),
+        stt=deepgram.STT(),
+        llm=gpt,
+        tts=elevenlabs.TTS(),
+        fnc_ctx=AssistantFnc(),
+    )
+
+    @assistant.on("will_synthesize_llm")
+    def _will_synthesize_llm(chat_ctx: llm.ChatContext, user_msg: llm.ChatMessage):
```

Comment:
this may be weird, I'm thinking about adding the chat_context to the constructor. But it may be confusing on whoever own the object

Reply:
how often is this called? Is it a single time before the llm is initialized? in that case, it'd be better to call the function directly instead of a callback.

Reply:
will followup in another PR for dynamic chat context